### DISCUSSION TITLE
Add option to change default microphone device

### DISCUSSION THREAD

**Discussion by sergeybruh** [File-New-Project/EarTrumpet#650] [Ideas]
Guys could you please add function of changing default microphone ?
Only one reason don't give me turn-off sound icon, I can't gent quick access to 
sound configuration form you app, to change microphone I would love you to add
there button to go to sonud devices setup or if is possible just add function of adding default device 
![image](https://cloud.githubusercontent.com/assets/13604069/21404475/0cc50810-c7ca-11e6-99b7-c9255b64e953.png)

![image](https://cloud.githubusercontent.com/assets/13604069/21404464/feff5e6a-c7c9-11e6-802c-d81475c12162.png)

Kind regards,
Sergey.

**Comment by Droyk** [comment]
+1 and plz add keyboard shortcut for them so that user can easily switch b/w them.

**Comment by riverar** [comment]
Changing the default mic device is interesting, will investigate that! Thanks.

**Comment by Kramins** [comment]
I would also request for Voice Gating support similar to VoiceMeeter.

**Comment by Justin-Credible** [comment]
In addition, it would be nice if there was a hotkey that would allow for toggling mute. My use case is toggling the mic when using Xbox Live Party chat (since the app doesn't have a push to talk and I don't want my mic recording the entire time, only when I want to speak). I currently use [MicMute](https://sourceforge.net/projects/micmute/) for this, but it isn't very reliable.

**Comment by ghostlyghastly** [comment]
It would be nice to have microphone with volume slider in there
my problem is that my microphone's volume setting gets changed - no idea by which software - and getting to the ui where i can check/reset this requires a lot of clicks and brain-processing of cluttered ui (window's sound settings)

**Reply by ghost** [reply]
same for me.. disabled it (allow applications to take exclusive control of this device) but apps like discord still can control my microphone.. even with their own mic control setting.

**Comment by vertigo220** [comment]
Any update on this, i.e. what the likelihood of it being added is? I'm constantly having to change the mic manually because Windows is apparently too stupid to figure it out on its own (which doesn't make sense, since it will switch to the headset for output but not for input), and it would be so much nicer to be able to right-click the ET icon and select the mic vs right-click > legacy > recording devices > select mic > set default > ok.

**Reply by ghost** [reply]
2022 and it seems it is still not here, currently goin' trough all discussions to see if it got announced to appear soon.. no luck till now~ :(

well there is a issue with someone having a micro icon there~ so i think riverar is currently working at it 
#1032

**Reply by riverar** [reply]
EarTrumpet is focused on managing audio output not input. However, we do have an addon framework and a sample addon that demonstrates microphone management. I'll see if I can find time to make sure this works in dev builds and get the addon documentation/repositories tested and published.

It's slow going as we have only have one active maintainer at this time (me).

**Reply by riverar** [reply]
Wanted to add: Adding it just to the right-click context menu might be a smaller easier task to complete. Would that be helpful?

**Reply by ghost** [reply]
> Wanted to add: Adding it just to the right-click context menu might be a smaller easier task to complete. Would that be helpful?

For some, maybe! 
for me? no!
I often change the Microphone Input level or my main microphone, so a simple (open eartrumpet and just slide the microphone slider to 90% or switching to another mic) done.

But as you said (EarTrumpet is focused on managing audio output not input)~
I'll respect your choice! Take your time please and if you don't want to add this, don't do it :D

I'm already really thankful that EarTrumpet exists, thank you for taking your time to develop helpful software!! <3

Windows completely sucks at creating reliable UI's.. at least the designer team whatever/whoever.. or just the higher ups? idk how it is in there :,D can you explain why they do such stupid things like hiding the microphone input level? (i'm asking because I saw youre work at the Windows Development Team :D) behind a few dozen clicks?

Even the worst Windows (Win 10) was way better than Windows 11 ._. well Win 11 has some nice features but the UIX? UI etc are really kinda mah sometimes weird.. shouldn't it be easy to explore and to find all the features Windows has!? Like they already had a damn volume mixer? what did they do? Hiding it behind a new less reliable Design that doesn't cant do what the volume mixer did?? gosh :(..

As I said!! Im really thankful for all your time you take for your software! please keep it up!

**Reply by felipecrs** [reply]
> Wanted to add: Adding it just to the right-click context menu might be a smaller easier task to complete. Would that be helpful?

But that's certainly a good first step which can be implemented independently of volume control. I would appreciate it a lot, and certainly others would do too.

:)

**Comment by vertigo220** [comment]
Those wanting this feature should check out SoundSwitch. It complements ET very well and, for me at least, solved my problem related to this issue and then some. IMO, the features of both should be built in to Windows. Then again, maybe not, because then it would be poorly done, buggy, updated constantly against users' wills (no means no, M$), and then removed for no reason.

**Reply by sambul13** [reply]
Still nothing was done despite promises to add mic switching to the app Hotkeys settings. When??? Adding it to Context Menu is not the right approach, since all hotkeys should be set in one place in the App Settings.

**Comment by Davoleo** [comment]
Hello @riverar ,
I'm thinking of implementing this feature in EarTrumpet,
I've been taking a look at the codebase recently and found much of the code structure to be already present inside the `EarTrumpet.Actions` addon, for example the `SetDefaultDeviceAction` uses the `Device` class that seems to support both recording and playback devices.
These actions also seem to be already implemented in a generic way inside the `ActionProcessor` class.
Is there a reason why none of the ViewModels for these actions are ever used?
also reading the technical docs inside the projects I understood that addons are meant to be compiled separately and included later, but I haven't seen any of the features mentioned inside that addon to be present in the app in general so I'm a bit confused by that, is 
 the `EarTrumpet.Actions` addon not present by default? if so why is it in the program's codebase at all? is it just to give an example of how you can implementa addons?

**Reply by riverar** [reply]
Hi @Davoleo, we have a separate private repository containing a handful of addons that we built to demonstrate our EarTrumpet Addon platform. We we're going to make that public, but due to lack of resources, and Windows developer platform instability, we have not yet shipped/enabled the add-on framework. (But managing microphones is one add-on we already have.) This is why you are finding functionality in the codebase that is not used or under-utilized.

Right now, my to-do list is, in order:
1. Complete .NET 8 transition
2. Rework packaging, distribution pipeline to support .NET 8 inbox delivery
3. Ship all that to `dev` users, slowly roll out to `master`
4. Maintain repository throughout the process

After this work is complete, I can consider taking on more community contributions and/or reconsider shipping our add-on platform.

**Reply by Davoleo** [reply]
I see, sounds cool, thanks for your efforts.
I wanted to work on this mostly because i find it very handy to be able to change default microphone the same way I do with the playback device. 
So something like it was mentioned here would be very useful to me https://github.com/File-New-Project/EarTrumpet/discussions/650#discussioncomment-2061205 as I don't need other advanced features ;
so I was working on adding a new context item separated from the rest of items called "Recording Devices" which then expands to a list of recording devices where you can select which one you want to be used as default.
Is it something I should keep in my fork just for me or should I create a PR for it? 
I'm aware that this would mean duplicating code to create a temporary solution for something coming down the line, so feel free to tell me if that's something you'd rather not have.

**Reply by riverar** [reply]
I say keep it and sit on it for a bit, if you don't mind. I'm completely overwhelmed with work at the moment but I did just recently finish the core transition to .NET 8. We now have x86, x64, and arm64 builds of EarTrumpet. I'm working on item 2 now, and as with everything in the Windows dev ecosystem, I hit a few bugs...
